File: app/static/passports.js
let activePassport = null;
pages.passports = async function () {
  const [products, sites] = await Promise.all([
    api("/api/passports"),
    api("/api/sites"),
  ]);
  return (
    head(
      "PRODUCT LIFECYCLE",
      "Product passports",
      "A persistent identity for each installed window or door.",
      user.role === "admin"
        ? '<div class="actions"><button data-passport-action="site">Add site</button><button class="primary" data-passport-action="create">Create passport</button></div>'
        : "",
    ) +
    '<div class="notice">Product details are supplied by your team. They are not verified manufacturer specifications. Lifecycle notes are shared within your company; linked inspection reports retain their existing access restrictions.</div>' +
    `<div class="cards">${products.length ? products.map((p) => `<article class="panel"><div class="eyebrow">${esc(p.product)}</div><h3>${esc(p.label)}</h3><p>${esc(sites.find((s) => s.id === p.site_id)?.name || "Site")}</p><p class="muted">${esc(p.manufacturer)} ${esc(p.system_name)}</p><p><small>Serial: ${esc(p.serial_number || "Not recorded")}</small></p><button data-passport-id="${p.id}">Open passport →</button></article>`).join("") : empty("No products registered yet", "An administrator can add a site and register the first product.")}</div>` +
    `<section class="panel" style="margin-top:24px"><h3>Registered sites</h3>${sites.length ? sites.map((s) => `<div class="visit"><b>${esc(s.name)}</b><p>${esc(s.address || "Address not recorded")}</p></div>`).join("") : "<p>No sites registered. Existing customers remain available when adding a site.</p>"}</section>`
  );
};
pages.passport = async function () {
  const d = await api("/api/passports/" + activePassport);
  const p = d.passport;
  const analysis = d.failure_analysis;
  const patterns = analysis.patterns
    .map(
      (item) =>
        `<div class="visit"><div class="actions"><h4>${esc(item.diagnosis)}</h4>${badge(item.signal)}</div><p>${item.cases} linked inspection${item.cases === 1 ? "" : "s"} · ${item.resolved} resolved · ${item.not_resolved} not resolved · ${item.outcome_not_recorded} without retained outcome</p><p>${item.original_snapshot_cases} based on immutable original diagnosis · ${item.repeat_visit_required} requiring another visit</p>${item.requires_review ? '<p class="notice">Repeated diagnosis with an unresolved or follow-up outcome. Engineer review recommended.</p>' : ""}</div>`,
    )
    .join("");
  return (
    head(
      "PRODUCT PASSPORT",
      p.label,
      d.site.name,
      '<button data-go="passports">All passports</button>',
    ) +
    `<section class="panel"><div class="report-meta">${[
      ["Product", p.product],
      ["Manufacturer", p.manufacturer],
      ["System", p.system_name],
      ["Serial number", p.serial_number],
      ["Site address", d.site.address],
      ["Passport ID", p.id],
    ]
      .map(
        ([k, v]) =>
          `<div><small>${k}</small><b style="overflow-wrap:anywhere">${esc(v || "Not recorded")}</b></div>`,
      )
      .join(
        "",
      )}</div><p class="muted">Team-supplied identification. Verify product applicability against controlled manufacturer evidence.</p><div class="actions"><button data-passport-action="event">Add lifecycle note</button><button data-passport-action="link">Link inspection</button></div></section>` +
    `<section class="panel" style="margin-top:24px"><div class="panel-head"><h3>Repeat-failure intelligence</h3>${badge(`${analysis.visible_linked_inspections} visible inspections`)}</div><p class="muted">${esc(analysis.method)}</p><p>${analysis.original_snapshot_cases} immutable original diagnoses · ${analysis.latest_outcomes} latest retained outcomes · ${analysis.omitted_without_diagnosis} records omitted without a diagnosis</p>${patterns || "<p>No diagnosis patterns are available yet.</p>"}</section>` +
    `<div class="grid" style="margin-top:24px"><section class="panel"><h3>Lifecycle history</h3><p class="muted">Entries are retained. Add a correction note to clarify an earlier entry.</p>${d.events.map((e) => `<div class="visit"><small>${esc(e.occurred_on)} · Recorded ${date(e.created_at)}</small><h4>${esc(e.kind)}</h4><p style="white-space:pre-wrap">${esc(e.note)}</p></div>`).join("")}</section><section class="panel"><h3>Linked inspections</h3>${d.inspections.length ? d.inspections.map((j) => `<div class="visit"><b>${esc(j.reference || "Inspection")}</b><p>${esc(j.outcome || "Outcome not recorded")}</p>${j.repair_outcome ? `<p><b>Latest repair:</b> ${esc(j.repair_outcome.payload.actual_repair)}</p><p><b>Final checks:</b> ${esc(j.repair_outcome.payload.verification_checks || "Not recorded in legacy feedback")}</p><small>Outcome revision ${j.repair_outcome.version} of ${j.outcome_revision_count} · ${j.repair_outcome.integrity_valid ? "Integrity checked" : "Integrity check failed"}</small>` : "<small>No retained repair outcome.</small>"}<p><button data-report="${j.id}">View report →</button></p></div>`).join("") : "<p>No linked reports available to your account.</p>"}</section></div>`
  );
};
document.addEventListener("click", async (e) => {
  const b = e.target.closest("button");
  if (!b) return;
  if (b.dataset.passportId) {
    activePassport = b.dataset.passportId;
    await go("passport");
    return;
  }
  const action = b.dataset.passportAction;
  if (!action) return;
  await busy(b, async () => {
    if (action === "site") {
      const customers = await api("/api/customers");
      if (!customers.length) {
        toast("Add a customer before registering its site.", true);
        return;
      }
      openModal(
        "Add site",
        `<form id="siteForm">${select("Customer", "customer_id", customers, "", "Select customer…")}${field("Site name", "name", "", "text", 'required maxlength="200"')}${area("Site address", "address")}<p class="error form-error" role="alert"></p><button class="primary">Save site</button></form>`,
      );
    }
    if (action === "create") {
      const sites = await api("/api/sites");
      if (!sites.length) {
        toast("Add a site before registering its products.", true);
        return;
      }
      openModal(
        "Create product passport",
        `<form id="passportCreate">${select("Site", "site_id", sites, "", "Select site…")}${field("Product location / label", "label", "", "text", 'required maxlength="200"')}${select("Product", "product", ["Window", "French Door", "Residential Door", "Bifold", "Sliding Door", "Tilt & Turn"])}${field("Manufacturer", "manufacturer")}${field("System", "system_name")}${field("Serial number", "serial_number")}<p class="error form-error" role="alert"></p><button class="primary">Create passport</button></form>`,
      );
    }
    if (action === "event")
      openModal(
        "Add lifecycle note",
        `<form id="passportEvent">${select("Event type", "kind", ["Manufacture note", "Installation note", "Service note", "Warranty note", "Correction note"])}${field("Event date", "occurred_on", new Date().toISOString().slice(0, 10), "date", "required")}${area("Note", "note", "", 'required maxlength="10000"')}<p class="muted">This note is retained and visible to your company. It does not certify installation, warranty coverage or authorise remedial work.</p><p class="error form-error" role="alert"></p><button class="primary">Save note</button></form>`,
      );
    if (action === "link") {
      const jobs = await api("/api/jobs");
      if (!jobs.length) {
        toast("Save an inspection before linking it.", true);
        return;
      }
      openModal(
        "Link an inspection",
        `<form id="passportLink">${select(
          "Inspection",
          "job_id",
          jobs.map((j) => ({
            id: j.id,
            name: (j.reference || "Inspection") + " · " + j.customer,
          })),
          "",
          "Select inspection…",
        )}<p class="muted">Confirm this inspection concerns this physical product. The link is retained and cannot be moved to another passport.</p><p class="error form-error" role="alert"></p><button class="primary">Link inspection</button></form>`,
      );
    }
  });
});
document.addEventListener("submit", async (e) => {
  const f = e.target;
  if (
    !["siteForm", "passportCreate", "passportEvent", "passportLink"].includes(
      f.id,
    )
  )
    return;
  e.preventDefault();
  const b = f.querySelector("button");
  if (b.disabled) return;
  await busy(b, async () => {
    try {
      const data = Object.fromEntries(new FormData(f));
      const endpoint = {
        siteForm: "/api/sites",
        passportCreate: "/api/passports",
        passportEvent: `/api/passports/${activePassport}/events`,
        passportLink: `/api/passports/${activePassport}/inspections`,
      }[f.id];
      const result = await send(endpoint, data);
      if (f.id === "passportCreate") activePassport = result.id;
      $("modal").close();
      await go(f.id === "siteForm" ? "passports" : "passport");
      toast("Saved");
    } catch (error) {
      f.querySelector(".form-error").textContent = error.message;
    }
  });
});

File: app/passports.py
"""Persistent company site/product identity and append-only lifecycle records."""
import uuid
from datetime import date, datetime
from typing import Literal
from fastapi import Depends, HTTPException
from pydantic import BaseModel, Field, field_validator
from sqlalchemy import String, Text, ForeignKey, DateTime, select
from sqlalchemy.orm import Mapped, mapped_column, Session
from sqlalchemy.exc import IntegrityError
from .db import Base, get_db
from .models import now, Customer, Job, WorkOrder
from .audit import log


class Site(Base):
    __tablename__ = "sites"
    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id"), index=True)
    customer_id: Mapped[str] = mapped_column(ForeignKey("customers.id"), index=True)
    name: Mapped[str] = mapped_column(String(200))
    address: Mapped[str] = mapped_column(Text, default="")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=now)


class ProductPassport(Base):
    __tablename__ = "product_passports"
    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id"), index=True)
    site_id: Mapped[str] = mapped_column(ForeignKey("sites.id"), index=True)
    label: Mapped[str] = mapped_column(String(200))
    product: Mapped[str] = mapped_column(String(100))
    manufacturer: Mapped[str] = mapped_column(String(180), default="")
    system_name: Mapped[str] = mapped_column(String(180), default="")
    serial_number: Mapped[str] = mapped_column(String(180), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=now)


class PassportInspection(Base):
    __tablename__ = "passport_inspections"
    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    passport_id: Mapped[str] = mapped_column(ForeignKey("product_passports.id"), index=True)
    job_id: Mapped[str] = mapped_column(ForeignKey("jobs.id"), unique=True, index=True)


class PassportEvent(Base):
    __tablename__ = "passport_events"
    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    passport_id: Mapped[str] = mapped_column(ForeignKey("product_passports.id"), index=True)
    recorded_by_id: Mapped[int] = mapped_column(ForeignKey("users.id"))
    kind: Mapped[str] = mapped_column(String(40))
    occurred_on: Mapped[str] = mapped_column(String(10))
    note: Mapped[str] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=now)


class Trimmed(BaseModel):
    @field_validator('*', mode='before')
    @classmethod
    def trim(cls, value):
        return value.strip() if isinstance(value, str) else value


class SiteIn(Trimmed):
    customer_id: str
    name: str = Field(min_length=1, max_length=200)
    address: str = ""


class PassportIn(Trimmed):
    site_id: str
    label: str = Field(min_length=1, max_length=200)
    product: str = Field(min_length=1, max_length=100)
    manufacturer: str = Field(default="", max_length=180)
    system_name: str = Field(default="", max_length=180)
    serial_number: str = Field(default="", max_length=180)


class EventIn(Trimmed):
    kind: Literal['Manufacture note', 'Installation note', 'Service note', 'Warranty note', 'Correction note']
    occurred_on: date
    note: str = Field(min_length=1, max_length=10000)


class LinkIn(BaseModel):
    job_id: str


def register(app, user_dep, check_job):
    def owned(db, model, identifier, user):
        record = db.get(model, identifier)
        if not record or record.company_id != user.company_id:
            raise HTTPException(404, "Record not found")
        return record

    def admin(user):
        if user.role != 'admin':
            raise HTTPException(403, "Company administrator required")

    @app.get('/api/sites')
    def sites(user=Depends(user_dep), db: Session=Depends(get_db)):
        return db.scalars(select(Site).where(Site.company_id==user.company_id).order_by(Site.name)).all()

    @app.post('/api/sites')
    def create_site(data:SiteIn, user=Depends(user_dep), db:Session=Depends(get_db)):
        admin(user); owned(db,Customer,data.customer_id,user)
        site=Site(id=str(uuid.uuid4()),company_id=user.company_id,**data.model_dump())
        db.add(site);log(db,user.company_id,user.id,'site.created','site',site.id)
        db.commit();db.refresh(site);return site

    @app.get('/api/passports')
    def passports(user=Depends(user_dep),db:Session=Depends(get_db)):
        return db.scalars(select(ProductPassport).where(ProductPassport.company_id==user.company_id).order_by(ProductPassport.created_at.desc())).all()

    @app.post('/api/passports')
    def create_passport(data:PassportIn,user=Depends(user_dep),db:Session=Depends(get_db)):
        admin(user);owned(db,Site,data.site_id,user)
        product=ProductPassport(id=str(uuid.uuid4()),company_id=user.company_id,**data.model_dump())
        db.add(product);db.flush()
        db.add(PassportEvent(id=str(uuid.uuid4()),passport_id=product.id,recorded_by_id=user.id,kind='Passport created',occurred_on=date.today().isoformat(),note='Product identity registered. Manufacturer details are user supplied, not verified specifications.'))
        log(db,user.company_id,user.id,'passport.created','passport',product.id)
        db.commit();db.refresh(product);return product

    @app.get('/api/passports/{identifier}')
    def detail(identifier:str,user=Depends(user_dep),db:Session=Depends(get_db)):
        from .outcomes import history as outcome_history, serialise as serialise_outcome
        from .repeat_failures import analyse as analyse_failures
        product=owned(db,ProductPassport,identifier,user)
        events=db.scalars(select(PassportEvent).where(PassportEvent.passport_id==identifier).order_by(PassportEvent.created_at,PassportEvent.id)).all()
        query=select(Job).join(PassportInspection,PassportInspection.job_id==Job.id).where(PassportInspection.passport_id==identifier,Job.company_id==user.company_id)
        if user.role!='admin': query=query.where(Job.engineer_id==user.id)
        jobs=db.scalars(query).all()
        inspections=[]
        for job in jobs:
            rows=outcome_history(db,job)
            latest=serialise_outcome(rows[-1]) if rows else None
            inspections.append({'id':job.id,'reference':job.reference,'outcome':job.outcome,'created_at':job.created_at,
                                'repair_outcome':latest,'outcome_revision_count':len(rows)})
        return {'passport':product,'site':owned(db,Site,product.site_id,user),'events':events,
                'inspections':inspections,'failure_analysis':analyse_failures(db,jobs)}

    @app.post('/api/passports/{identifier}/events')
    def add_event(identifier:str,data:EventIn,user=Depends(user_dep),db:Session=Depends(get_db)):
        owned(db,ProductPassport,identifier,user)
        record=PassportEvent(id=str(uuid.uuid4()),passport_id=identifier,recorded_by_id=user.id,kind=data.kind,occurred_on=data.occurred_on.isoformat(),note=data.note)
        db.add(record);log(db,user.company_id,user.id,'passport.event_added','passport',identifier,{'event_id':record.id})
        db.commit();db.refresh(record);return record

    @app.post('/api/passports/{identifier}/inspections')
    def link(identifier:str,data:LinkIn,user=Depends(user_dep),db:Session=Depends(get_db)):
        product=owned(db,ProductPassport,identifier,user)
        job=db.get(Job,data.job_id);check_job(job,user)
        site=owned(db,Site,product.site_id,user)
        if job.customer_id and job.customer_id!=site.customer_id:
            raise HTTPException(409,'Passport customer conflicts with the inspection customer')
        linked_customers=set(db.scalars(select(WorkOrder.customer_id).where(
            WorkOrder.company_id==user.company_id,WorkOrder.job_id==job.id,
            WorkOrder.customer_id.is_not(None))).all())
        if any(customer_id!=site.customer_id for customer_id in linked_customers):
            raise HTTPException(409,'Passport customer conflicts with a linked work order')
        prior=db.scalar(select(PassportInspection).where(PassportInspection.job_id==job.id))
        if prior:
            if prior.passport_id==identifier:return {'ok':True}
            raise HTTPException(409,'This inspection is already linked to another product passport')
        db.add(PassportInspection(id=str(uuid.uuid4()),passport_id=identifier,job_id=job.id))
        db.add(PassportEvent(id=str(uuid.uuid4()),passport_id=identifier,recorded_by_id=user.id,kind='Inspection linked',occurred_on=date.today().isoformat(),note='An inspection was linked. Its report remains subject to assigned-engineer access.'))
        log(db,user.company_id,user.id,'passport.inspection_linked','passport',identifier,{'job_id':job.id})
        try:db.commit()
        except IntegrityError:
            db.rollback();raise HTTPException(409,'Inspection link changed; refresh and try again')
        return {'ok':True}
